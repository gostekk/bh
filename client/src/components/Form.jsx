import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Material-ui
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import withStyles from '@material-ui/core/styles/withStyles';


const propTypes = {
  classes: PropTypes.shape({
    container: PropTypes.string.isRequired,
    textField: PropTypes.string.isRequired,
    submit: PropTypes.string.isRequired,
  }).isRequired,
  addNewPerson: PropTypes.func.isRequired,
  resetSuccess: PropTypes.func.isRequired,
  success: PropTypes.bool.isRequired,
  errors: PropTypes.shape({
    firstname: PropTypes.string,
    lastname: PropTypes.string,
    email: PropTypes.string,
    eventdate: PropTypes.string,
  }).isRequired,
};

const styles = theme => ({
  container: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  textField: {
    marginLeft: theme.spacing.unit,
    marginRight: theme.spacing.unit,
  },
  submit: {
    marginTop: theme.spacing.unit * 3,
  },
});

export class Form extends Component {
  constructor(props) {
    super(props);

    this.state = {
      firstname: '',
      lastname: '',
      email: '',
      eventdate: '',
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidUpdate(prevProps) {
    const { success, resetSuccess } = this.props;

    if (success && !prevProps.success) {
      this.resetState();
      resetSuccess();
    }
  }

  resetState() {
    this.setState({
      firstname: '',
      lastname: '',
      email: '',
      eventdate: '',
    });
  }

  handleChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  handleSubmit(e) {
    e.preventDefault();
    const {
      firstname,
      lastname,
      email,
      eventdate,
    } = this.state;
    const { addNewPerson } = this.props;

    const newPerson = {
      firstname,
      lastname,
      email,
      eventdate,
    };

    addNewPerson(newPerson);
  }

  render() {
    const {
      firstname,
      lastname,
      email,
      eventdate,
    } = this.state;
    const { classes, errors } = this.props;

    return (
      <form className={classes.container} noValidate autoComplete="off" onSubmit={this.handleSubmit}>
        <TextField
          required
          fullWidth
          id="firstname"
          label="Firstname"
          className={classes.textField}
          value={firstname}
          onChange={this.handleChange}
          name="firstname"
          error={!!errors.firstname}
          helperText={errors.firstname}
          margin="normal"
          variant="outlined"
        />
        <TextField
          required
          fullWidth
          id="lastname"
          label="Lastname"
          className={classes.textField}
          value={lastname}
          onChange={this.handleChange}
          name="lastname"
          error={!!errors.lastname}
          helperText={errors.lastname}
          margin="normal"
          variant="outlined"
        />
        <TextField
          required
          fullWidth
          id="email"
          label="email"
          className={classes.textField}
          value={email}
          onChange={this.handleChange}
          type="email"
          name="email"
          autoComplete="email"
          error={!!errors.email}
          helperText={errors.email}
          margin="normal"
          variant="outlined"
        />
        <TextField
          required
          fullWidth
          id="eventdate"
          label="Event Date"
          type="date"
          value={eventdate}
          onChange={this.handleChange}
          className={classes.textField}
          error={!!errors.eventdate}
          helperText={errors.eventdate}
          name="eventdate"
          InputLabelProps={{
            shrink: true,
          }}
          margin="normal"
          variant="outlined"
        />
        <Button
          id="submit"
          type="submit"
          fullWidth
          variant="raised"
          color="primary"
          className={classes.submit}
        >
          Submit
        </Button>
      </form>
    );
  }
}

Form.propTypes = propTypes;

export default withStyles(styles)(Form);
