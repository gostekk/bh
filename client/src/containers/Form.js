import { connect } from 'react-redux';
import { addNewPerson, resetSuccess } from '../actions/form';
import Form from '../components/Form.jsx';

export const mapStateToProps = state => ({
  success: state.form.success,
  errors: state.form.errors,
});

export const mapDispatchToProps = dispatch => ({
  addNewPerson: person => dispatch(addNewPerson(person)),
  resetSuccess: () => dispatch(resetSuccess()),
});

const FormContainer = connect(mapStateToProps, mapDispatchToProps)(Form);
export default FormContainer;
