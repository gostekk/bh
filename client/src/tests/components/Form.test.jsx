import React from 'react';
import { shallow } from 'enzyme';
import Form from '../../components/Form.jsx';
import TextField from '@material-ui/core/TextField';

describe('Form Component', () => {
  let wrapper;
  let props;
  
  beforeEach(() => {
    props = {
      addNewPerson: jest.fn(),
      resetSuccess: jest.fn(),
      success: false,
      errors: {},
    };

    wrapper = shallow(<Form {...props} />).dive();
  });

  it('should render Form component and match to snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should click button and call addNewPerson function', () => {
    wrapper.find('form').simulate('submit', { preventDefault() {} });
    expect(props.addNewPerson).toHaveBeenCalled();
  });

  it('should reset state of the component after true in success prop', () => {
    wrapper.find(TextField).at(0).props().onChange({ target: { value: 'test', name: 'firstname' } });
    expect(wrapper.state().firstname).toEqual('test');
    expect(wrapper.props().success).toBeFalsy();
    wrapper.setProps({ success: true });
    expect(wrapper.state().firstname).toEqual('');
  });
});
