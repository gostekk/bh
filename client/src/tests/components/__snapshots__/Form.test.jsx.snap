// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component should render Form component and match to snapshot 1`] = `
<form
  autoComplete="off"
  className="Form-container-1"
  noValidate={true}
  onSubmit={[Function]}
>
  <TextField
    className="Form-textField-2"
    error={false}
    fullWidth={true}
    id="firstname"
    label="Firstname"
    margin="normal"
    name="firstname"
    onChange={[Function]}
    required={true}
    select={false}
    value=""
    variant="outlined"
  />
  <TextField
    className="Form-textField-2"
    error={false}
    fullWidth={true}
    id="lastname"
    label="Lastname"
    margin="normal"
    name="lastname"
    onChange={[Function]}
    required={true}
    select={false}
    value=""
    variant="outlined"
  />
  <TextField
    autoComplete="email"
    className="Form-textField-2"
    error={false}
    fullWidth={true}
    id="email"
    label="email"
    margin="normal"
    name="email"
    onChange={[Function]}
    required={true}
    select={false}
    type="email"
    value=""
    variant="outlined"
  />
  <TextField
    InputLabelProps={
      Object {
        "shrink": true,
      }
    }
    className="Form-textField-2"
    error={false}
    fullWidth={true}
    id="eventdate"
    label="Event Date"
    margin="normal"
    name="eventdate"
    onChange={[Function]}
    required={true}
    select={false}
    type="date"
    value=""
    variant="outlined"
  />
  <WithStyles(Button)
    className="Form-submit-3"
    color="primary"
    fullWidth={true}
    id="submit"
    type="submit"
    variant="raised"
  >
    Submit
  </WithStyles(Button)>
</form>
`;
